commands/up.go

```diff
@@ -64,20 +77,133 @@ func preRunUp(cmd *cobra.Command, args []string) error {
 }
 
 func upHandler(cmd *cobra.Command, args []string) error {
-	if err := runBuild(cmd, args); err != nil {
+	handler := upRunner(cmd, args)
+	if err := handler(); err != nil {
 		return err
 	}
-	fmt.Println()
-	if !skipPush {
-		if err := runPush(cmd, args); err != nil {
+
+	if watch {
+		logger := log.New(os.Stderr, "", log.LstdFlags)
+		logger.SetFlags(log.Ldate | log.Ltime | log.LUTC)
+
+		watcher, err := fsnotify.NewWatcher()
+		if err != nil {
 			return err
 		}
-		fmt.Println()
-	}
-	if !skipDeploy {
-		if err := runDeploy(cmd, args); err != nil {
+		defer watcher.Close()
+
+		patterns, err := ignorePatterns()
+		if err != nil {
+			return err
+		}
+
+		if err = filepath.Walk("./", func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			for _, pattern := range patterns {
+				match, err := filepath.Match(pattern, path)
+				if err != nil {
+					return err
+				}
+
+				if match {
+					return filepath.SkipDir
+				}
+			}
+
+			if info.IsDir() {
+				return watcher.Add(path)
+			}
+			return nil
+		}); err != nil {
 			return err
 		}
+
+		signalChannel := make(chan os.Signal, 1)
+		signal.Notify(signalChannel, os.Interrupt, syscall.SIGTERM)
+
+		d := debounce.New(2 * time.Second)
+
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return fmt.Errorf("watcher's Events channel is closed")
+				}
+
+				if event.Op == fsnotify.Write {
+					d(func() {
+						if err := handler(); err != nil {
```

Comment:
@alexellis i have been thinking about command cancelling a bit more, in short, it will require a refactor in at least two project. 

To have the ability to cancel concurrent runs, we would need to pass and manage a Context through all of the build layers so that we can pass it to the actual build command. 

This obviously requires a lot of changes in this project, but that isn't enough. 

Back when you first wrote the builder you opted to wrap the `exec.Cmd` with https://github.com/alexellis/go-execute (which hasn't really seen or needed any changes in the last 3 years) 

Most importantly go-execute does not support passing a context or cancelling a command. The ideal situation would be to use https://pkg.go.dev/os/exec#CommandContext so that a context can be passed down to the command, allowing cancellation. 

Either go-execute needs an update, which can then be used here, or faas-cli needs to refactor back to the stdlib

Reply:
I think you're right about the context cancellation. We should definitely update go-execute for that. It hasn't seen changes because the API and requirements have been stable.

Reply:
@NikhilSharmaWe and @alexellis i think we have to decide if this PR needs to include a change to allow context propagation to the build command or if we want to do that in a future change. 

go-execute is only really used once (that I can see) in this project. So, I would be in favor of just reverting to the stdlib here. Which would allow this PR to immediately update and add the required context propagation. Or updating once go-execute has context support. 

In either case, once we are building and cancelling a context based on the signal, the only other change i would like to see is on [L149](https://github.com/openfaas/faas-cli/pull/962/files#diff-5165889054d8170eac2603b4631a88d1addad1739d20c60d20a51949bffbaff5R149) when the watcher close can also watch the `ctx.Done()` channel for consistency.

Reply:
Actually thinking about this more, we should just merge this and fix the context cancellation in a follow up.  The current state won't hurt anyone because it is an opt in feature

Reply:
Are you volunteering for that @LucasRoesler ?